### DISCUSSION TITLE
Please, explain me why forward call 2 times update() function

### DISCUSSION THREAD

**Discussion by PososikTeam** [Lightning-AI/torchmetrics#953] [General]
Please, explain me why forward method call 2 times update() function. In Metric class :
```
@torch.jit.unused
    def forward(self, *args: Any, **kwargs: Any) -> Any:
        """Automatically calls ``update()``.

        Returns the metric value over inputs if ``compute_on_step`` is True.
        """
        # add current step
        if self._is_synced:
            raise TorchMetricsUserError(
                "The Metric shouldn't be synced when performing ``update``. "
                "HINT: Did you forget to call ``unsync`` ?."
            )

        with torch.no_grad():
            self.update(*args, **kwargs)

        if self.compute_on_step:
            self._to_sync = self.dist_sync_on_step
            # skip restore cache operation from compute as cache is stored below.
            self._should_unsync = False

            # save context before switch
            cache = {attr: getattr(self, attr) for attr in self._defaults}

            # call reset, update, compute, on single batch
            self.reset()
            self.update(*args, **kwargs)
            self._forward_cache = self.compute()
```

**Comment by SkafteNicki** [comment]
The first call updates the global statistics of the metric (accumulation over multiple batches) and the second call computes the metric on the current batch. That said, we are aware this is inefficient and we are trying to get rid of the double call: https://github.com/PyTorchLightning/metrics/pull/984. This should hopefully be merged before the next release.
